theano/tensor/nnet/nnet.py

```diff
@@ -786,16 +934,102 @@ def local_logsoftmax_grad(node):
                        subtensor.AdvancedIncSubtensor))):
         # get parameters from unoptimized op
         sm = node.inputs[0].owner.inputs[1]
+        axis = sm.owner.op.axis
         # sm_input = node.inputs[1].owner.inputs[0]
         grads = node.inputs[0].owner.inputs[0]
-        if grads.broadcastable[1] and not sm.broadcastable[1]:
-            grads = tensor.alloc(grads, grads.shape[0], sm.shape[1])
-        ret = grads - tensor.sum(grads, axis=1, keepdims=True) * sm
+        if grads.broadcastable[axis] and not sm.broadcastable[axis]:
+            grads = tensor.alloc(grads, grads.shape[0], sm.shape[axis])
+        ret = grads - tensor.sum(grads, axis=axis, keepdims=True) * sm
         ret.tag.values_eq_approx = values_eq_approx_remove_nan
         copy_stack_trace(node.outputs[0], ret)
         return [ret]
 
 
+@opt.register_specialize('fast_compile')
+@gof.local_optimizer([SoftmaxGrad])
+def local_indexing_logsoftmax_grad(node):
+    # One case is supported:
+    # 1. AdvancedIncSubtensor(
+    #           zeros_like(softmax(x)),
+    #           -out_grad / AdvancedSubtensor(softmax(x), arange(y.shape[0]), y),
+    #           arange(y.shape[0]),
+    #           y)
+    #   which arises from the gradient of log(softmax(x)[arange(y.shape[0]), y])
+    list_classes_IncSubtensor = ((IncSubtensor, AdvancedIncSubtensor1, AdvancedIncSubtensor))
+    list_classes_Subtensor = ((Subtensor, AdvancedSubtensor1, AdvancedSubtensor))
+    if not (isinstance(node.op, SoftmaxGrad)):
+        return
+
+    sm = None
+    try:
+        d_sm, sm = node.inputs
+    except Exception:
+        return
+
+    if not((sm is not None) and sm.owner and isinstance(sm.owner.op, Softmax)):
+        return
+
+    if d_sm.owner and isinstance(d_sm.owner.op, list_classes_IncSubtensor):
+        try:
+            incr = d_sm.owner.inputs[1]
+            subtensor_inputs, = d_sm.owner.inputs[2:]
+        except:
+            return
```

Comment:
Can you remove the try/except? I don't think it is right to silence that error without understanding the possible issues.